### ISSUE TITLE
Mask land cells in the Loss calculation

### ISSUE THREAD

**Issue by PaulSpence** [PaulSpence/OM2-emulator#13]
from om2_autencoder.ipynb

# Loss function and optimizer
criterion = nn.L1Loss()

This loss function needs to ignore land cells in ocean data (similar to Ryan's partial convolution function).

**Comment by rmholmes** [comment]
@PaulSpence the below is a cut-down version of my training loop with the masked loss. Hopefully it is adaptable. Let me know if you have questions (I can send you all the code if you want, there is just a lot in there that is not relevant).

`hr_batch` is the truth, `pred` is the predictor, `mask_hr` is the land-sea mask (as an xarray). Note that the `mask_b` is for the partial convolutions, not the loss mask. The working code is around `loss_mask_b` and the `F.mse_loss`

```
import torch.nn as nn
import torch.nn.functional as F

def train_epoch(model, loader, optimizer, device, mask, static_preds={}, loss_mask=None):
    model.train()
    running_loss = 0.0
    batches = 0
    for hr_bilin_batch, hr_batch in loader:
        hr_bilin_batch = hr_bilin_batch.to(device)
        hr_batch = hr_batch.to(device)
        optimizer.zero_grad()
        B, C, H, W = hr_batch.shape
        mask_b = mask.expand(B, 1, H, W)

        x = hr_bilin_batch
        for key in static_preds.keys():
            x = torch.cat(
                [x, static_preds[key].expand(B, 1, H, W)], 
                dim=1)
        
        pred = model(x, mask_b)

        # Masked loss:
        if loss_mask is None:
            # default: full-domain mask of ones
            loss_mask_b = torch.ones((B, 1, H, W), device=device)
        else:
            # expand user mask to batch size
            loss_mask_b = loss_mask.expand(B, 1, H, W)
        
        loss = F.mse_loss(pred, hr_batch, reduction='none')
        loss = (loss * loss_mask_b).sum() / loss_mask_b.sum()
        
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        batches += 1
    return running_loss / max(batches, 1)


# Main code that is relevant:

    # Loss mask definition:
    lmask = mask_hr.values
    lmask = torch.tensor(lmask.values).float()   # shape (H,W)
    lmask = lmask.unsqueeze(0).unsqueeze(0)      # shape (1,1,H,W) so it broadcasts well
    lmask = lmask.to(device)

   # Training loop call:
        train_mse = train_epoch(model, train_loader, optimizer, device, mask, static_preds=spreds, loss_mask=lmask)
        train_losses.append(train_mse)

**Comment by rmholmes** [comment]
I have implemented this in https://github.com/PaulSpence/OM2-emulator/tree/Pconv Needs checking.

**Comment by rmholmes** [comment]
Done.
